RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,750 @@
+
+* [x] [Approved in principle](https://github.com/fsharp/fslang-suggestions/issues/579#event-1345361104) & [prioritised](https://github.com/fsharp/fslang-suggestions/issues/579#event-1501977428)
+* [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/579)
+* [x] [Discussion](https://github.com/fsharp/fslang-design/issues/335)
+* [ ] Implementation: [In progress](https://github.com/Microsoft/visualfsharp/pull/5696)
+
+# Summary
+[summary]: #summary
+
+Extend computation expressions to support applicative functors via a new `let! ... and! ... return ...` syntax.
+
+# Motivation
+[motivation]: #motivation
+
+Applicative functors (or just "applicatives", for short) have been growing in popularity as a way to build applications and model certain domains over the last decade or so, since McBride and Paterson published [Applicative Programming with Effects](http://www.staff.city.ac.uk/~ross/papers/Applicative.html). Applicatives are now reaching a level of popularity within the community that supporting them with a convenient and readable syntax, as we do for monads, makes sense.
+
+With applicative computation expressions, we can write more computations than before (there are more contexts which meet the requirements for applicative computation expressions than the existing monadic ones), and we can write more efficient computations (the requirements of applicatives rule out needing to support some potentially expensive operations).
+
+For example, [Pauan points out](https://github.com/fsharp/fslang-suggestions/issues/579#issuecomment-310799948) that we can write a computation expression for `Observable`s that [avoids unnecessary resubscriptions](https://github.com/fsharp/fslang-suggestions/issues/579#issuecomment-310854419):
+
+```fsharp
+// Outputs a + b, which is recomputed every time foo or bar outputs a new value,
+// avoiding any unnecessary resubscriptions
+observable {
+  let! a = foo
+  and! b = bar
+  return a + b
+}
+```
+
+## Why Applicatives?
+
+[Applicative functors](https://en.wikipedia.org/wiki/Applicative_functor) sit, on the spectrum of flexibility vs. predictability, somewhere between [functors](https://en.wikipedia.org/wiki/Functor#Computer_implementations) (i.e. types which support `Map`), and [monads](https://en.wikipedia.org/wiki/Monad_(functional_programming)) (i.e. types which support `Bind`, which currently underpin computation expressions).
+
+If we consider `Bind : M<'T> * ('T -> M<'U>) -> M<'U>`, we can see that the second element of the input is a function that requires a value to create the resulting "wrapped value". This means the argument to `Bind` has the power to completely change the context of the result based on the value seen (e.g. to create and destroy `Observable` subscriptions), but it also means that the expression builder can predict much less about what the given function will decide to do, and hence has fewer outcomes that it can rule out and potentially optimise away.
```

Comment:
It's worth noting that while the documentation specifies this `M` remains consistent throughout, you can write valid computation expressions (especially type extensions) that replace an input `M` with another type, e.g. type extensions for `AsyncBuilder` that allow binding `Task<'T>` or `Task`.

Reply:
Interesting thought. I'd been working on the basis that putting in anything but a consistent `M` was going off-piste somewhat, so I planned to skirt over it. Perhaps it is worth covering that after all.

Reply:
Another example would be allowing `Bind` (or `Apply` in this case`) against a `null` or `Nullable<'T>` in an `OptionBuilder`.

Reply:
I highlighted some of this flexibility in https://github.com/panesofglass/computation-expressions-workshop

Reply:
The [MSDN docs](https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/computation-expressions#creating-a-new-type-of-computation-expression) talk about "typical signatures" of what is being desugared. I suppose that is a bit different from when I am coming from in this RFC. I don't think basing this change in the established literature and theory is a bad thing, but mentioning that you _can_ use other signatures I think probably is important, as you say.

Reply:
I've added a small section dedicated to clarifying that other method signatures are possible.